docs/flip/1009-optimizer-api.md

```diff
@@ -0,0 +1,471 @@
+parts of the parameter tree with different optimizers:
+
+```python
+biases_traversal = flax.optim.ModelParamTraversal(
+    lambda path, _: path.endswith('/bias'))
+not_biases_traversal = flax.optim.ModelParamTraversal(
+    lambda path, _: not path.endswith('/bias'))
+
+optimizer_def = flax.optim.MultiOptimizer(
+    (biases_traversal, flax.optim.GradientDescent(learning_rate=0.1)),
+    (not_biases_traversal, flax.optim.GradientDescent(learning_rate=0.05)),
+)
+```
+
+Note how we first define a traversal that selects parameters based on their
+path (which is the concatenation of module scopes and variable name), and then
+create a `MultiOptimizer` that binds a different optimizer for each of these
+separate traversals.
+
+Optax has recently implemented `optax.masked()` that can be used for specifying
+gradient transformations that only applied to a subset of the gradients:
+
+```python
+def flattened_traversal(fn):
+  def mask(data):
+    flat = traverse_util.flatten_dict(data)
+    return traverse_util.unflatten_dict({k: fn(k, v) for k, v in flat.items()})
+  return mask
+
+tx = optax.chain(
+    optax.masked(optax.sgd(learning_rate=0.1),
+                 mask=flattened_traversal(lambda path, _: path[-1] == 'bias')),
+    optax.masked(optax.sgd(learning_rate=0.05),
+                 mask=flattened_traversal(lambda path, _: path[-1] != 'bias')),
+)
+```
+
+## Train State
+[Train State]: #train-state
+
+In Flax it is common to hand around a `TrainState` object that can then be
+used for checkpointing. This simplifies the above [Optax training step] a bit by
+reducing the number of arguments and getting rid of the `static_argnums`.
+
+We can define an `Optimizer` dataclass that wraps the common pattern of updating
+the optimizer state and parameters, and keeps `params` that are to be optimized
+separate from `model_state` that is not touched by the optimizer:
+
+```python
+class Optimizer(flax.struct.PyTreeNode):
+  params: flax.core.FrozenDict[str, Any]
+  model_state: flax.core.FrozenDict[str, Any]
+  tx: optax.GradientTransformation = flax.struct.field(pytree_node=False)
+  opt_state: optax.OptState
+
+  def update(self, grads, new_model_state):
+    updates, new_opt_state = self.tx.update(
+        grads, self.opt_state, self.params)
+    new_params = optax.apply_updates(self.params, updates)
+    return self.replace(
+        params=new_params,
+        opt_state=new_opt_state,
+        model_state=new_model_state,
+    )
+
+  @classmethod
+  def create(cls, variables, tx):
+    model_state, params = variables.pop('params')
+    opt_state = tx.init(params)
+    return cls(
+        params=params,
+        model_state = model_state,
+        tx=tx,
+        opt_state=opt_state,
+    )
+```
+
+Users can then compose their own data class containing this optimizer, the
+model's `.apply()` function and arbitrary additional data:
+
+```python
+class TrainState(flax.struct.PyTreeNode):
```

Comment:
What's the benefit of having separate `Optimizer` and `TrainState` classes? (instead of putting everything in `TrainState`?)

One downside of the current proposal, is something I've heard from confused users -- when you load from a checkpoint, why do you need to go through an "optimizer object" to get the parameters if all you're doing is making a prediction?

Reply:
In an earlier version I had an implementation that moved everything at the top level in the dataclass. But jheek and n2cholas convinced me in [above comment](https://github.com/google/flax/pull/1011#discussion_r593151131) to use composition instead.

Reply:
Sorry if I was unclear before, I prefer having everything in `TrainState` as well. I was just in favour of splitting up `params` and `model_state` _within_ the `TrainState` class.

Reply:
Got it, sorry for mischaracterising your comment!

That still leaves jheek's suggestion to use composition. Both his arguments about dataclasses not behaving well under inheritance, and the possibility to have multiple optimizers/models, speak in favour of composition.

While I see Avital's point that users might not expect to load the model weights "from the optimizer", I think this is just a new pattern to get used to, and I would weight the arguments about inheritance & having multiple optimizers/models more heavily.

Pinging @jheek if he wants to add something .

Reply:
I am proposing neither inheritance nor composition -- a single `TrainState` dataclass.

Reply:
I think there are 2 viable options:
- Let users roll their own "flat" TrainState where they manually track params, model_state, opt_state, and the optimizer (as a meta). 
- Create an Optimizer dataclass that users can use for their own TrainState (actually usually it replaces TrainState all together because these 4 fields are the only state).

I prefer the second option considering how much discussion we had already about the optimizer pattern. The bookkeeping itself is non-trivial and it might not be obvious to users that params & model_state need to be split or that the Optax function needs to be a meta field. The abstraction itself doesn't get in the way and users could also opt-out of it easily by "inlining" these fields into their own TrainState object. Eventually we might have more utilities that make it easier to use like an optimize method that takes a loss_fn and a gradient reducer or a property that gives the variables by combining params & modle_state.

Reply:
Yes indeed I misunderstood the intent that the proposal is that `TrainState` is user-defined, while `Optimizer` is part of Flax. I now better understand the rationale.

Perhaps I am stating a small preference for not calling it `Optimizer` then... We can't call it `OptimizerState` because that's an Optax concept. I still think this confuses many users as it's neither an optimizer nor the optimizer state. It's more like "model and optimizer metadata"... 

It's like the good old saying about only two hard things in software engineering...

I am thinking out loud here: Why wouldn't we call what's proposed as `Optimizer` here, instead, `TrainState`? Or `BaseTrainState`? Or `CommonTrainState`? It's not really the optimizer state, as it also holds parameters -- I don't think most folks would intuitively expect an "optimizer instance" to hold parameters.

Then it goes back to being confusing that users will then go ahead and define their own `TrainState` that holds a `CommonTrainState`... I'd like to better understand the concrete downsides of using inheritence, because this seems pretty similar to how we define `Module` and users extend it, which they can then call `MyModule` or whatever they want... To me this seems quite similar. I'll chat with @jheek offline for a few minutes to understand his objection.

Reply:
Reworked this part:

- As per Avital's suggestions, switched back to inheritance (which should work just fine with dataclasses as long as they don't have optional fields), and keeping everything at the root level. This will allow users to access parameters simply via `state.params`, not going through the `state.optimizer` attribute.
- Concerning the naming, going simply for `TrainState`. Since this will live in a different module, users can later subclass via `class TrainState(flax.training.TrainState)`
- Also added `apply_fn` and handling of `step` to the default implementation because `apply_fn` will be required in most cases and I think it's often useful to have the step number in the train state (anyways it shouldn't harm when it's not needed).
- The current solution works without modification for the simple state-less case.
- It's easy and safe to subclass `TrainState` if the user wants to store/update mutable model state or other data.

Reply:
I really like this latest version. It's short, understandable, really easy to extend (just add more dataclass attributes for additional state collections), you can probably easily override `update` if you want to do some modifications to parameters on updating if for some reason it's not supported in Optax, and you can also simply fork the whole TrainState class or not use it at all.

Reply:
I agree, the latest version looks great!